File: antargis/trunk/src/ant_hero.cc
/*
 * Copyright (c) 2005 by David Kamphausen. All rights reserved.
 *
 * ant_hero.cc
 * by David Kamphausen (david.kamphausen@web.de)
 *
 * The "Antargis" project, including all files needed to compile it,
 * is free software; you can redistribute it and/or use it and/or modify it
 * under the terms of the GNU General Public License as published
 * by the Free Software Foundation; either version 2 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program.
 */

#include "ant_hero.h"
#include "ant_man.h"
#include "map.h"
#include "jobs.h"


AntHero::AntHero():typeID(0)
{}

/************************************************************************
* AntHero
************************************************************************/
AntHero::AntHero(const Pos2D &p,int pTypeID,const std::string &pName):AntBoss(p),typeID(pTypeID),mName(pName)
{}
AntHero::~AntHero()
{
  std::set
    <AntMan*>::iterator i=mMen.begin();
  for(;i!=mMen.end();i++)
    (*i)->discard(this);

  getMap()->killHero(this);
}
void AntHero::updateSurface()
  {
    std::ostringstream os;
    if(typeID==0)
      os<<"hero1dl";
    else
      os<<"hero2dl";
    VoxelImage *im=new VoxelImage(os.str());//imageCache()->getImage(os.str());
    im->setPosition(mPos);
    //im->setCenter(Pos2D(100,150)+Pos2D(0,64));
    im->setVirtualY(40);
    im->setName(mName);
    setSurface(im);
  }

int AntHero::getTypeID() const
  {
    return typeID;
  }


void AntHero::fightHero(AntHero *h)
{
  CTRACE;
  // myselfs attacks other hero
  setJob(new FightJob(0,h));

  // get other's men and other hero and put them all into a vector
  std::set
    <AntMan*> otherMen=h->mMen;
  std::vector<AntEntity*> otherList;

  std::set
    <AntMan*>::iterator i=otherMen.begin();
  for(;i!=otherMen.end();i++)
    otherList.push_back(*i);
  otherList.push_back(h); // push other hero

  std::random_shuffle(otherList.begin(),otherList.end()); // shuffle all

  std::vector<AntEntity*>::iterator k=otherList.begin();
  // assign from front to back all the entities for fighting
  for(i=mMen.begin();i!=mMen.end();i++)
    {
      (*i)->setJob(new FightJob(0,*k));

      k++;
      if(k==otherList.end())
        k=otherList.begin(); // simply restart, if end is hit
    }


}

void AntHero::goTo(int prio,const Pos2D &pos)
{
  std::set
    <AntMan*>::iterator i;
  for(i=mMen.begin();i!=mMen.end();i++)
    (*i)->jobFinished();
  setJob(new MoveJob(prio,pos));
}

/*
void AntHero::signUp(AntMan *man)
{
  mMen.insert(man);
}


void AntHero::discard(AntMan *man)
{
  mMen.erase(man);
  }*/

float AntHero::calcTroupStrength() const
  {
    float s=getEnergy();
    std::set
      <AntMan*>::iterator i;
    for(i=mMen.begin();i!=mMen.end();i++)
      s+=(*i)->getEnergy();
    return s;
  }

AntHero *AntHero::fights()
{
  if(mJob)
    {
      FightJob *f=dynamic_cast<FightJob*>(mJob);
      if(f)
        {
          return dynamic_cast<AntHero*>(f->getTarget());
        }
    }
  return 0;
}

void AntHero::saveXML(xmlpp::Node &node) const
{
  AntEntity::saveXML(node);
  node.set("typeID",toString(typeID));
  node.set("name",mName);
}
void AntHero::loadXML(const xmlpp::Node &node)
{
  AntEntity::loadXML(node);
  typeID=toInt(node.get("typeID"));
  mName=node.get("name");
}


void AntHero::gotFight(AntEntity *e)
{
  if(!fights())
    {
      AntHero *h=e->getHero();
      
      if(h)
            fightHero(h);
    }
}

Pos2D AntHero::getFormation(AntMan *m) const
{
  size_t count=mMen.size();
  
  // if(count>20) FIXME: do second circle
  
  std::set<AntMan*>::const_iterator j=mMen.begin();
  size_t c=0;
  for(;j!=mMen.end() && *j!=m;j++,c++)
    ;
  if(j==mMen.end())
    return Pos2D(0,0);
  else
    {
      
      if(hasJob())
	{
	  MoveJob *j=dynamic_cast<MoveJob*>(mJob);
	  if(j)
	    {
	      Pos2D dir=j->getDirection(this);
	      Pos2D normal=dir.normal();
	      Pos2D t=dir*(-1)*(c/6)*32+normal*((c%6)-2.5)*16;
	      return t;
	    }
	}
      float angle=float(c)/float(count)*M_PI*2.0;
      return Pos2D(sin(angle)*64,cos(angle)*64);
    }
  
}

AntHero *AntHero::getHero()
{
  return this;
}
float AntHero::getHealSpeed() const
{
  if(hasJob())
    return mHealSpeed*0.2;
  else
    return mHealSpeed;
}
std::string AntHero::xmlName() const
{
  return "antHero";
}

File: antargis/trunk/ruby/ant_hero.rb
#!/usr/bin/env ruby
#
# Copyright (c) 2005 by David Kamphausen. All rights reserved.
#
# ant_hero.rb
# by David Kamphausen (david.kamphausen@web.de)
#
# The "Antargis" project, including all files needed to compile it,
# is free software; you can redistribute it and/or use it and/or modify it
# under the terms of the GNU General Public License as published
# by the Free Software Foundation; either version 2 of the License,
# or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
# You should have received a copy of the GNU General Public
# License along with this program.
#

#!/usr/bin/ruby

require 'ents.rb'
require 'ant_hljobs.rb'
require 'ant_boss.rb'
require 'ant_fire.rb'
require 'ant_ring.rb'
require 'ant_manbase.rb'

class AntHero<AntBoss
	include AntManBase

	attr_reader :meshState, :dead
	def initialize
		super
		@men.push(self)
		setProvide("hero",true)
		@appearance="hero"
		getMap.setLight(self)
		setMinimapColor(AGColor.new(0xFF,0,0))
		@meshStand="stand"
		self.age=(getRand*20+20)
		setHunger(0.006)
	end
	def setAppearance(a)
		@appearance=a
	end
	
	def saveXML(node)
		super(node)
		node.set("appearance",@appearance.to_s)
	end
	def loadXML(node)
		super(node)
		@appearance=node.get("appearance")
		if @appearance==""
			@appearance="hero"
		end
		@portrait=node.get("portrait")

		if node.get("primary")=="true"
			#setMoraleHealSpeed=1000
			setMoraleFlag(false)
		end
	end
	
	def noHLJob
		puts "noHLJob"
		if @player
			@player.assignJob(self)
			#stopFireSound
		else
			# no player , so simply rest - to infinity (or at least 5 seconds)
			newHLRestJob(5)
		end
	end
	
	def startFireSound
		if not @fireSound
			dputs "STARTING FIRE"
			@fireSound=playLoopSoundGlobal(self,"fire",getPos2D,0.4)
			dputs @fireSound
		end
	end	
	def stopFireSound
		if @fireSound
			dputs "STOPPED"
			dputs @job
			if @job.class==AntHeroRestJob
				#raise "bla"
			end
			stopLoopSound(self)
			#getSoundManager.stopChannel(@fireSound)
			@fireSound=nil
		end
	end
	
	def newHLRestJob(time)
		@job=nil
		setFire(true)
		super(time)
		assignJob2All
	end
	
	def assignJob(man)
		checkHLJobEnd(man)
	end	
	def moveHome(man)
		if @formation.nil?
			raise "formation is nil!"
		end
		pos=@formation.getPosition(man)
# 		pos=getSitFormation(man)
		if (man.getPos2D-pos).length>1
			man.newMoveJob(0,pos,0)
		else
			man.newRestJob(2)
		end
	end
	
	
	def newHLMoveJob(prio,pos,dist)
		@job=AntHeroMoveJob.new(self,prio,pos,dist)
		assignJob2All
	end
	def newHLRecruitJob(target)
		@job=AntHeroRecruitJob.new(self,target,$buttonPanel.getAggression)
		assignJob2All
	end
	def newHLConstructJob(target)
		@job=AntHeroConstructJob.new(self,target,$buttonPanel.getAggression)
		assignJob2All
	end
	def newHLTakeFoodJob(target)
		@job=AntHeroTakeJob.new(self,target,$buttonPanel.getAggression,"food")
		assignJob2All
	end
	def newHLTakeWeaponJob(target)
		@job=AntHeroTakeJob.new(self,target,$buttonPanel.getAggression,"weapon")
		assignJob2All
	end
	def newHLFightJob(target)
		@job=AntHeroFightJob.new(self,target)
		assignJob2All
	end
	def newHLFightAnimalJob(target)
		@job=AntHeroFightAnimalJob.new(self,target)
		assignJob2All
	end
	def newHLBuildJob(pos,type)
		@job=AntHeroBuildJob.new(self,pos,type)
		assignJob2All
	end
	
	def newHLDismissJob()
		@job=nil
		agg=$buttonPanel.getAggression
		men=@men.select{|m|not m.is_a?(AntHero)} # exclude hero
		c=(men.length-1)*agg/3
			
		men=men[0..c]
		men.each{|m|
			m.setNoBoss
			@men.delete(m)
			m.delJob
		}
		getMap.eventHLDismissed(self)
	end
	
	def upgradeMan(man)
		# copy properties through xml-node	
		xmlnode=Node.new("test")
		man.saveXML(xmlnode)
		
		nhero=AntHero.new
		nhero.loadXML(xmlnode)
		removeMan(man)
		getMap.removeEntity(man)
		getMap.insertEntity(nhero)
		#getMap.endChange # commit
		nhero.newRestJob(0)
		newHLRestJob(1)
		#nhero.setName("Sir "+nhero.getName)
		man.setName("xxxxxxxx") # reset name - so there's some distinction
		nhero.setOwner(self)
		nhero.experience=0
	end
	
	
	def setFire(flag)
		puts "setFire #{flag}"
		if flag
			if getPos3D.z>0 # won't start fire in water!!
				if not @fire
					@fire=AntFire.new(getPos3D+AGVector3.new(0.7,0,0))
					getMap.insertEntity(@fire)
				end
				startFireSound
			end
		else
			#raise 1
			if @fire
				@fire.disable
				@fire=false
			end
			stopFireSound
		end
	end

	def eventAttacked(by)
		puts "eventAttacked #{by}"
		newHLDefendJob(by)
	end
	
	def assignJob2All
		super
		dputs "ASSIGNJOB"
		if @job.class!=AntHeroRestJob
			setFire(false)
		end
		doEvent(:newJobAssigned)
	end

	def setMeshState(name)

		if name=="row"
			name="walk"
		end

		@origMeshState=name
		name=checkOnWater(name)
		puts "FIXME: implement setMeshState(.)"
		@meshState=name
		dir=getDirection
		case name
			when "row"
		 		setMesh(Mesh.new(getMap.getScene,getMeshData("data/models/hero_lp.ant2",0.08,"data/textures/models/hero_lp.png"),AGVector4.new(0,0,0,0),0))
				#setMesh(AnimMesh.new(getMap.getScene,getAnimMeshData("data/models/man_e_walk.anim")))
				#getFirstMesh.setAnimation("sit")
				mesh=Mesh.new(getMap.getScene,getBoatMeshData,AGVector4.new(0,0,0),0)
				addMesh(mesh,AGVector3.new(0,0,0))

			when "dead"
		 		setMesh(Mesh.new(getMap.getScene,getMeshData("data/models/grave.ant2",0.2,"data/textures/models/grave3.png"),AGVector4.new(0,0,0,0),0))
			else
		 		setMesh(Mesh.new(getMap.getScene,getMeshData("data/models/hero_lp.ant2",0.08,"data/textures/models/hero_lp.png"),AGVector4.new(0,0,0,0),0))
		end
		setDirection(dir)
		setupRing

	end



 	def setupMesh
		setMeshState("normal")
	end
	def getRing
		makeRingMesh
	end

	def getImage
		@portrait||=""
		r=nil

		filename=""
		if @portrait.length!=0
			puts "port!=0"
			filename=@portrait
			#r=getTextureCache.get(@portrait)
		else
			puts "port==0"
			#r=getTextureCache.get("data/gui/portraits/#{getName}.png")
			filename="data/gui/portraits/#{getName}.png"
		end
		if fileExists(filename)
			r=getTextureCache.get(filename)
		else
			r=getTextureCache.get("data/gui/portraits/none.png")
		end

		puts "getImage-ok"
		return r
	end
	def getDescription
		nameage=_("This is {1}. He is {2} years old.",getName,age)
		support=nil
		if getMen.length==1
			support=_("Nobody supports him.")
		else
			support=_("{1} men support him.",getMen.length-1)
		end
		nameage+" "+support
	end

	def spreadResources
		# FIXME
	end

	def eventDie
		super
		@job=nil
		if @player
			@player.remove(self)
		end
		getMap.eventHeroDied(self)
	end

end

